sdk/python/lib/pulumi/x/automation/events.py

```diff
@@ -0,0 +1,537 @@
+                 metadata: StepEventMetadata,
+                 planning: Optional[bool] = None):
+        self.metadata = metadata
+        self.planning = planning
+
+    @classmethod
+    def from_json(cls, data: dict) -> 'ResOutputsEvent':
+        return cls(**data)
+
+
+class ResOpFailedEvent(BaseEvent):
+    """
+    ResOpFailedEvent is emitted when a resource operation fails. Typically a DiagnosticEvent is
+    emitted before this event, indicating the root cause of the error.
+    """
+    def __init__(self,
+                 metadata: StepEventMetadata,
+                 status: int,
+                 steps: int):
+        self.metadata = metadata
+        self.status = status
+        self.steps = steps
+
+    @classmethod
+    def from_json(cls, data: dict) -> 'ResOpFailedEvent':
+        return cls(**data)
+
+
+class EngineEvent(BaseEvent):
+    """
+    EngineEvent describes a Pulumi engine event, such as a change to a resource or diagnostic
+    message. EngineEvent is a discriminated union of all possible event types, and exactly one
+    field will be non-nil.
+
+    Attributes
+    ----------
+    sequence: str
+        Sequence is a unique, and monotonically increasing number for each engine event sent to the
+        Pulumi Service. Since events may be sent concurrently, and/or delayed via network routing,
+        the sequence number is to ensure events can be placed into a total ordering.
+        - No two events can have the same sequence number.
+        - Events with a lower sequence number must have been emitted before those with a higher
+          sequence number.
+    timestamp: int
+        Timestamp is a Unix timestamp (seconds) of when the event was emitted.
+    """
+    def __init__(self,
+                 sequence: int,
+                 timestamp: int,
+                 cancel_event: Optional[CancelEvent] = None,
+                 stdout_event: Optional[StdoutEngineEvent] = None,
+                 diagnostic_event: Optional[DiagnosticEvent] = None,
+                 prelude_event: Optional[PreludeEvent] = None,
+                 summary_event: Optional[SummaryEvent] = None,
+                 resource_pre_event: Optional[ResourcePreEvent] = None,
+                 res_outputs_event: Optional[ResOutputsEvent] = None,
+                 res_op_failed_event: Optional[ResOpFailedEvent] = None,
+                 policy_event: Optional[PolicyEvent] = None):
+        self.sequence = sequence
+        self.timestamp = timestamp
+        self.cancel_event = cancel_event
+        self.stdout_event = stdout_event
+        self.diagnostic_event = diagnostic_event
+        self.prelude_event = prelude_event
+        self.summary_event = summary_event
+        self.resource_pre_event = resource_pre_event
+        self.res_outputs_event = res_outputs_event
+        self.res_op_failed_event = res_op_failed_event
+        self.policy_event = policy_event
+
+    @classmethod
+    def from_json(cls, data: dict) -> 'EngineEvent':
```

Comment:
This is works and gives precise control. I wonder if in the future we'll want matching to_json. There seems to be something out there that can attempt metaprogramming based name conversion so we don't have to spell it out, though ofc we lose some control:

https://pypi.org/project/dataclasses-json/

Reply:
We have thus far avoided using dataclasses because they were introduced in python 3.7 and even though there is a [backport](https://pypi.org/project/dataclasses/), @justinvp hasn't been too into the idea of taking the dependency.

It's also not always as straightforward as pure snake case to camel case, as in the case of [PolicyPacks](https://github.com/pulumi/pulumi/pull/6527/files#diff-1c8278368b2a8ab8d889dc9ffa979f0abfdf1c646782cefbe0b22f8443d4fa55R244)

Reply:
Right, sounds like that one is out of reach. OK let's move on.